src/nix-env/nix-env.cc

```diff
@@ -911,36 +911,44 @@ static void queryJSON(Globals & globals, vector<DrvInfo> & elems, bool printOutP
 {
     JSONObject topObj(cout, true);
     for (auto & i : elems) {
-        JSONObject pkgObj = topObj.object(i.attrPath);
-
-        auto drvName = DrvName(i.queryName());
-        pkgObj.attr("name", drvName.fullName);
-        pkgObj.attr("pname", drvName.name);
-        pkgObj.attr("version", drvName.version);
-        pkgObj.attr("system", i.querySystem());
-
-        if (printOutPath) {
-            DrvInfo::Outputs outputs = i.queryOutputs();
-            JSONObject outputObj = pkgObj.object("outputs");
-            for (auto & j : outputs) {
-                outputObj.attr(j.first, j.second);
+        try {
+            if (i.hasFailed()) continue;
+
+            JSONObject pkgObj = topObj.object(i.attrPath);
+
+            auto drvName = DrvName(i.queryName());
+            pkgObj.attr("name", drvName.fullName);
+            pkgObj.attr("pname", drvName.name);
+            pkgObj.attr("version", drvName.version);
+            pkgObj.attr("system", i.querySystem());
+
+            if (printOutPath) {
+                DrvInfo::Outputs outputs = i.queryOutputs();
+                JSONObject outputObj = pkgObj.object("outputs");
+                for (auto & j : outputs) {
+                    outputObj.attr(j.first, j.second);
+                }
             }
-        }
 
-        if (printMeta) {
-            JSONObject metaObj = pkgObj.object("meta");
-            StringSet metaNames = i.queryMetaNames();
-            for (auto & j : metaNames) {
-                auto placeholder = metaObj.placeholder(j);
-                Value * v = i.queryMeta(j);
-                if (!v) {
-                    printError("derivation '%s' has invalid meta attribute '%s'", i.queryName(), j);
-                    placeholder.write(nullptr);
-                } else {
-                    PathSet context;
-                    printValueAsJSON(*globals.state, true, *v, noPos, placeholder, context);
+            if (printMeta) {
+                JSONObject metaObj = pkgObj.object("meta");
+                StringSet metaNames = i.queryMetaNames();
+                for (auto & j : metaNames) {
+                    auto placeholder = metaObj.placeholder(j);
+                    Value * v = i.queryMeta(j);
+                    if (!v) {
+                        printError("derivation '%s' has invalid meta attribute '%s'", i.queryName(), j);
+                        placeholder.write(nullptr);
+                    } else {
+                        PathSet context;
+                        printValueAsJSON(*globals.state, true, *v, noPos, placeholder, context);
+                    }
                 }
             }
+        } catch (AssertionError & e) {
+            printMsg(lvlTalkative, "skipping derivation named '%1%' which gives an assertion failure", i.queryName());
+        } catch (Error & e) {
+            printMsg(lvlError, "skipping derivation named '%1%' which gives an error '%2%'", i.queryName(), e.msg());
```

Comment:
The table format `throws` here instead.

Reply:
Yeah we should not ignore arbitrary errors.

Reply:
Fixed in latest commit.